src/odm_sharing/private/cons.py

```diff
@@ -23,37 +34,186 @@ def _write_table_to_db(db: sa.engine.Engine, table: str, df: pd.DataFrame
     df.to_sql(table, db, index=False, if_exists='replace')
 
 
+def _datasets_to_db(datasets: Dict[str, pd.DataFrame]) -> sa.engine.Engine:
+    '''creates an in-memory db and writes the datasets as tables'''
+    db = _create_memory_db()
+    for table, df in datasets.items():
+        _write_table_to_db(db, table, df)
+    return db
+
+
+def _get_csv_dataset_bool_cols(df: pd.DataFrame) -> Set[str]:
+    '''Finds bool columns.
+
+    All casings of the words true and false are included.
+
+    If a normal (non-NA non-bool) text value is encountered, then the column is
+    assumed to be a normal text column (and not a bool column) and is ignored.
+
+    :returns: a set of column names
+    '''
+    # XXX: Strings are stored with dtype=object since they are Python objects.
+    # No other Python objects can be imported from a spreadsheet, so it should
+    # always be `str`.
+    BOOL_VALS = ['FALSE', 'TRUE']
+    NA_VALS = ['', 'NA']
+    result: Set[str] = set()
+    for col in df:
+
+        # check column type
+        if df[col].dtype == bool:
+            result.add(str(col))
+            continue
+        if df[col].dtype != object:  # not str
+            continue
+
+        # check string value: NA, bool, or other
+        for val in df[col]:
+            norm_val = val.upper()
+            if norm_val in NA_VALS:  # NA
+                continue
+            if norm_val in BOOL_VALS:  # bool
+                result.add(str(col))
+            break
+    return result
+
+
 def _connect_csv(path: str) -> Connection:
     '''copies file data to in-memory db
 
     :raises OSError:'''
+    # XXX: NA-values are not normalized to avoid mutating user data (#31)
     logging.info('importing csv file')
     table = Path(path).stem
-    db = _create_memory_db()
-    df = pd.read_csv(path)
-    _write_table_to_db(db, table, df)
-    return db
+    df = pd.read_csv(path, na_filter=False)
+    bool_cols = _get_csv_dataset_bool_cols(df)
+
+    # normalize bool (string) values to 0/1
+    #
+    # XXX: Bool columns with NA-values still represent bools as FALSE/TRUE.
+    # This needs to be normalized to 0/1, like normal bool columns, so that all
```

Comment:
normal bool columns don't need any pre-processing since they're already binary, so they're not normalized either. String columns with bool values are normalized here to use the binary values 0/1 so that we can query them in the same way as with the normal bool columns (by comparing against 0/1).

Reply:
I meant the normalisation to 0/1, that conversion only happens when we put them into the sqlite db right? pandas still stores them as the Python True/False

Reply:
I'm making it 0/1 before inserting to the db, otherwise it may be inserted as string as well ('false'/'true'), which wouldn't have the same value as normal bool columns and fail the query filter.